include/deal.II/lac/trilinos_tpetra_vector.h

```diff
@@ -474,20 +761,220 @@ namespace LinearAlgebra
        * CommunicationPattern for the communication between the
        * source_stored_elements IndexSet and the current vector.
        */
-      std::shared_ptr<const TpetraWrappers::CommunicationPattern>
+      Teuchos::RCP<const TpetraWrappers::CommunicationPattern>
         tpetra_comm_pattern;
+
+      // Make the reference class a friend.
+      friend class internal::VectorReference<Number>;
     };
 
 
+    /* ------------------------- Inline functions ---------------------- */
+
     template <typename Number>
     inline bool
     Vector<Number>::has_ghost_elements() const
     {
       return false;
     }
+
+
+
+    template <typename Number>
+    inline void
+    Vector<Number>::add(const std::vector<size_type>      &indices,
+                        const std::vector<TrilinosScalar> &values)
+    {
+      // if we have ghost values, do not allow
+      // writing to this vector at all.
+      AssertDimension(indices.size(), values.size());
+
+      add(indices.size(), indices.data(), values.data());
+    }
+
+
+
+    template <typename Number>
+    inline void
+    Vector<Number>::add(const size_type  n_elements,
+                        const size_type *indices,
+                        const Number    *values)
+    {
+#  if DEAL_II_TRILINOS_VERSION_GTE(13, 2, 0)
```

Comment:
I hope we can require a more recent version soonish...

Reply:
I hope so, too!

I just added the support for older Trilinos versions after my last pull request caused some problems in the regression test.